### PR TITLE
[SPARK-35183][SQL] Use transformAllExpressions in CombineConcats

### PR DESCRIPTION
**Pull request by sigmod**
### What changes were proposed in this pull request?

Use transformAllExpressions instead of transformExpressionsDown in CombineConcats. The latter only transforms the root plan node.

### Why are the changes needed?

It allows CombineConcats to cover more cases where `concat` are not in the root plan node.

### How was this patch tested?

Unit test. The updated tests would fail without the code change.

### CODE REVIEW

**Comment by sigmod** [comment]
@maropu @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 1]
```diff
@@ -506,6 +506,18 @@ class ExplainSuite extends ExplainSuiteHelper with DisableAdaptiveExecutionSuite
       checkKeywordsExistsInExplain(df2, keywords = "[key1=value1, KEY2=VALUE2]")
     }
   }
+
+  test("SPARK-35183: CombineConcats should call transformAllExpressions") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:510]: It's a bit weird to check the plan using explain result. If we want to have an end-to-end test for this optimization, let's check `sql(...).queryExecution.optimizedPlan` directly.

Normally we only write UT to test optimizer changes, so I'm fine to not have an end-to-end test here.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala (hunk 1 above)
Comment (sigmod) [reply sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala]: Done. Reverted. 

(The explain test was copied and modified from another [one](https://github.com/apache/spark/blob/f208d808819dae38739fa9acda9510a4b9763dde/sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala#L151-L159) testing the original rule.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by maropu** [comment]
late lgtm.

**Pull request closed without merging**